winpr/libwinpr/sspi/Negotiate/negotiate.c

```diff
@@ -1407,9 +1407,9 @@ static SECURITY_STATUS SEC_ENTRY negotiate_AcquireCredentialsHandleW(
 		const SecPkg* pkg = MechTable[i].pkg;
 		cred->mech = &MechTable[i];
 
-		if (!kerberos && _tcscmp(pkg->name, KERBEROS_SSP_NAME) == 0)
+		if (!kerberos && _tcsncmp(pkg->name, KERBEROS_SSP_NAME, ARRAYSIZE(KERBEROS_SSP_NAME)) == 0)
```

Comment:
while it adds a little overhead (aka explicitly write the length as argument) it prevents reintroducing these in code paths where we do not want such functions to be used (most of `FreeRDP`)

since it is impossible to distinguish these valid instances from the other ones the only feasible way is to just eradicate them completely and bail error once they are reintroduced.
(and we´re talking compile time overhead here)

Reply:
I don't understand your point,  `pkg->name` will always be maintained by us as it's an internal structure, and `KERBEROS_SSP_NAME` is constant. There's no possible misuse in the future. 
More generally, all the len checking string functions pointed by this PR can be useful when it's external data that we may not trust (and we have a way to have max length), but that doesn't seem like a good case when it's for some of our internal data.

Always the same remark with trust within a code units and checks: if the point is "you never know, that doesn't hurt to add that checks", then what are the rules to add the checks ? If we start taking measures like if we don't trust our own internal data, then we'll end up with checking everything possible everywhere. We'll have a meta-FreeRDP inside FreeRDP that checks that FreeRDP does what is expected !

Note: it's not only pushing on the stack the length argument, this n argument also have to be check at every loop turn in `_tcsncmp` (so check for EOS and for n)

Reply:
@hardening point being there are functions from the standard library more error prone than others.
There is a reason such function are deprecated/are warned about by a lot of checkers.

anyway, the changes to kerberos have grown too big for this pr, will split it out in a separate one.